visible.lua

```diff
@@ -0,0 +1,156 @@
+local LIBRARY_PATH = (...):match("(.-)[^%.]+$")
+local vector = require(LIBRARY_PATH .. "vector")
+local visible = {}
+
+
+local function comp(a, b, center)
+  if a.start == b.start then
+    if a.stop == nil then return false end
+    if b.stop == nil then return true end
+    return vector.alignment(a.start, a.stop, b.stop) > 0
+  end
+  return vector.polar_lt(a.start, b.start, center)
+end
+
+function visible.polygon(segments, center)
+  -- generic visibility function
+  -- return concave (frequently) polygon
+  if center == nil then center = vector.zero end
+  local endpoints = {}
+  do
+    local next_endpoints = {}
+    local missing_stops = {}
+    for i = 1, #segments, 4 do
```

Comment:
In each loop here you clearly do a distinct operation. Name them and create local functions for them. it makes the code more modular, easier to build on and easier to read.

Reply:
For example, you might name this one collect_endpoints